File: timus/1671.py
import sys
import gc

gc.disable()

class Node:
    def __init__(self):
        self.parent = None
        self.rank   = 0

class Edge:
    def __init__(self, fr, to):
        self.fr     = fr
        self.to     = to
        self.isInQ  = False

class UnionFind:

    def __init__(self, size):
        self.connectedComponentsCount = size
        self.roots = {}

    def find_root(self, node):
        while node != node.parent:
            node = node.parent
        return node.parent

    def union(self, node1, node2):
        self.connectedComponentsCount -= 1
        if node1.rank > node2.rank:
            node2.parent = node1
        else:
            node1.parent = node2
            if node1.rank == node2.rank:
                node2.rank += 1

    def process(self, edge):
        fromRoot = self.find_root(edge.fr)
        toRoot = self.find_root(edge.to)
        if fromRoot != toRoot:
            self.union(fromRoot, toRoot)

N, M = map(int, sys.stdin.readline().split())

nodes = []
for i in range(N):
    nodes.append(Node())
    nodes[-1].parent = nodes[-1]

edges = []
for i in range(M):
    fr, to = map(int, sys.stdin.readline().split())
    edges.append(Edge(nodes[fr-1], nodes[to-1]))

Q = int(sys.stdin.readline())

removes = map(int, sys.stdin.readline().split())

qedges = []
qedges = [edges[remove-1] for remove in removes]
for qedge in qedges:
    qedge.isInQ = True

uf = UnionFind(N)

for edge in edges:
    if not edge.isInQ:
        uf.process(edge)

result = []
for qedge in reversed(qedges):
    result.append(uf.connectedComponentsCount)
    uf.process(qedge)

for r in reversed(result):
    sys.stdout.write(str(r) + " ")
sys.stdout.write('\n')

File: timus/1671.cpp
#include <stdio.h>
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <set>
#include <list>
#include <algorithm>

using namespace std;

struct TNode {
    TNode *Parent;
    int Rank;
    TNode() : Parent(0) , Rank(0) { }
};

struct TEdge {
    TNode *From;
    TNode *To;
    bool IsInQ;
    TEdge() : From(0) , To(0) , IsInQ(false) { }
};

struct TDisjointSet {
    int ConnectedComponentsCount;

    TDisjointSet(int size) {
        ConnectedComponentsCount = size;
    }

    TNode* FindSetRoot(TNode *node) {
        if (node != node->Parent)
            node->Parent = FindSetRoot(node->Parent);
        return node->Parent;
    }

    void UnionRoots(TNode *r1, TNode *r2) {
        ConnectedComponentsCount--;
        if (r1->Rank > r2->Rank)
            r2->Parent = r1;
        else {
            r1->Parent = r2;
            if (r1->Rank == r2->Rank)
                (r2->Rank)++;
        }
    }

    void Process(TEdge &edge) {
        TNode *fromRoot = FindSetRoot(edge.From);
        TNode *toRoot = FindSetRoot(edge.To);
        if (fromRoot != toRoot)
            UnionRoots(fromRoot, toRoot);
    }
};

int main() {

    int N, M, Q;

    scanf("%d %d\n", &N, &M);

    vector<TNode> nodes(N);
    for (int i = 0; i < N; i++)
        nodes[i].Parent = &nodes[i];

    vector<TEdge> edges(M);
    int from, to;
    for (int i = 0; i < M; ++i) {
        scanf("%d %d\n", &from, &to);
        TEdge &edge = edges[i];
        edge.From = &(nodes[from-1]);
        edge.To = &(nodes[to-1]);
    }

    scanf("%d\n", &Q);
    vector<TEdge*> qedges(Q);
    int qnum;
    for (int i = 0; i < Q; ++i) {
        scanf("%d", &qnum); 
        TEdge &qedge = edges[qnum-1];
        qedge.IsInQ = true;
        qedges[i] = &qedge;
    }

    TDisjointSet ds(N);

    for (vector<TEdge>::iterator iter = edges.begin(), end = edges.end(); iter != end; ++iter) {
        TEdge &edge = *iter;
        if (edge.IsInQ)
            continue;
        ds.Process(edge);
    }

    vector<int> result;
    for (vector<TEdge*>::reverse_iterator iter = qedges.rbegin(), end = qedges.rend(); iter != end; ++iter) {
        TEdge &edge = **iter;
        result.push_back(ds.ConnectedComponentsCount);
        ds.Process(edge);
    }

    for (vector<int>::const_reverse_iterator iter = result.rbegin(), end = result.rend(); iter != end; ++iter)
        printf("%d ", *iter);
    printf("\n");

    return 0;
}
